File: packages/core/src/policies/scope.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Adam Campbell

/**
 * DefaultScopePolicy — flat-membership scope authorization.
 *
 * Decides whether the requestor's credential carries the scope the action
 * requires. A missing scope is a hard deny.
 *
 * Ported from Solid# `solid-backend/middleware/auth.py:376-422` and
 * `solid-backend/models/cli_api_key.py:178-182` (has_scope).
 * Conforms to Themis RFC v0 § 5.3.
 *
 * Decision rule (first match wins):
 *
 *   1. requestor.is_super_admin                        -> allow
 *   2. requestor.scopes includes '*'                   -> allow
 *   3. requestor.kind === 'user' && scopes is empty    -> allow  (session user)
 *   4. required_scope resolves to null                 -> allow  (no scope needed)
 *   5. requestor.scopes includes required_scope        -> allow
 *   6. otherwise                                       -> deny
 *
 * `required_scope` is taken from `action.required_scope` first. If absent,
 * the policy consults its registered rule table (path + method, or predicate)
 * to resolve one. If no rule matches, the action passes scope policy (rule 4).
 *
 * Flat membership is the default per RFC v0: `pages:write` does NOT imply
 * `pages:read`. Wildcards beyond `*` are opt-in extensions (not in core).
 */

import type {
  IPolicy,
  IPolicyContext,
  IPolicyDecision,
  IScopePolicy,
  ScopeRuleMatcher,
} from '../types.js';

interface ScopeRule {
  matcher: ScopeRuleMatcher;
  required_scope: string;
}

export class DefaultScopePolicy implements IScopePolicy, IPolicy {
  readonly name = 'scope' as const;
  private readonly rules: ScopeRule[] = [];

  async evaluate(ctx: IPolicyContext): Promise<IPolicyDecision> {
    // 1. Super admin bypass
    if (ctx.requestor.is_super_admin === true) {
      return { kind: 'allow', policy: this.name };
    }

    // 2. Wildcard scope — granted full access
    if (ctx.requestor.scopes.includes('*')) {
      return { kind: 'allow', policy: this.name };
    }

    // 3. Session users (kind='user', no scopes) bypass scope checks
    if (ctx.requestor.kind === 'user' && ctx.requestor.scopes.length === 0) {
      return { kind: 'allow', policy: this.name };
    }

    // 4. Resolve the required scope
    const required = ctx.action.required_scope ?? this.resolveRequired(ctx);
    if (!required) {
      return { kind: 'allow', policy: this.name };
    }

    // 5. Flat membership check
    if (ctx.requestor.scopes.includes(required)) {
      return { kind: 'allow', policy: this.name };
    }

    // 6. Deny
    return {
      kind: 'deny',
      policy: this.name,
      reason: 'missing_scope',
      message: `Missing required scope '${required}'.`,
      detail: {
        required,
        held: Array.from(ctx.requestor.scopes),
      },
    };
  }

  addRule(matcher: ScopeRuleMatcher, required_scope: string): void {
    this.rules.push({ matcher, required_scope });
  }

  /** Return every distinct scope mentioned in any registered rule. */
  listScopes(): readonly string[] {
    const seen = new Set<string>();
    for (const r of this.rules) seen.add(r.required_scope);
    return Array.from(seen);
  }

  private resolveRequired(ctx: IPolicyContext): string | null {
    for (const { matcher, required_scope } of this.rules) {
      if (typeof matcher === 'function') {
        if (matcher(ctx)) return required_scope;
      } else {
        const method = (ctx.action.metadata?.['method'] as string | undefined) ?? '';
        const path = (ctx.action.metadata?.['path'] as string | undefined) ?? '';
        if (method.toUpperCase() !== matcher.method.toUpperCase()) continue;
        if (typeof matcher.path_pattern === 'string') {
          if (path === matcher.path_pattern) return required_scope;
        } else if (matcher.path_pattern.test(path)) {
          return required_scope;
        }
      }
    }
    return null;
  }
}

File: python/src/themis/policies/scope.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Adam Campbell
"""DefaultScopePolicy — flat-membership scope authorization (RFC v0 § 5.3).

Decision rule, first match wins:
  1. requestor.is_super_admin                    → allow
  2. '*' in requestor.scopes                     → allow
  3. requestor.kind == 'user' and no scopes      → allow  (session user)
  4. required scope resolves to None             → allow
  5. required scope in requestor.scopes          → allow
  6. otherwise                                   → deny
``required_scope`` comes from the action first, else from the registered rule
table (method + path string/regex, or a predicate). Flat membership: no
implied hierarchy; wildcards beyond '*' are extensions, not core.
"""
from __future__ import annotations

import re
from typing import Callable, Optional, Pattern, Union

from ..types import Allow, Decision, Deny, PolicyContext

Matcher = Union[tuple[str, Union[str, Pattern[str]]], Callable[[PolicyContext], bool]]


class DefaultScopePolicy:
    name = "scope"

    def __init__(self) -> None:
        self._rules: list[tuple[Matcher, str]] = []

    def evaluate(self, ctx: PolicyContext) -> Decision:
        r = ctx.requestor
        if r.is_super_admin is True:
            return Allow(policy=self.name)
        if "*" in r.scopes:
            return Allow(policy=self.name)
        if r.kind == "user" and len(r.scopes) == 0:
            return Allow(policy=self.name)
        required = ctx.action.required_scope or self._resolve_required(ctx)
        if not required:
            return Allow(policy=self.name)
        if required in r.scopes:
            return Allow(policy=self.name)
        return Deny(
            policy=self.name,
            reason="missing_scope",
            message=f"Missing required scope '{required}'.",
            detail={"required": required, "held": list(r.scopes)},
        )

    def add_rule(self, matcher: Matcher, required_scope: str) -> None:
        self._rules.append((matcher, required_scope))

    def list_scopes(self) -> list[str]:
        seen: dict[str, None] = {}
        for _, s in self._rules:
            seen.setdefault(s, None)
        return list(seen)

    def _resolve_required(self, ctx: PolicyContext) -> Optional[str]:
        meta = ctx.action.metadata or {}
        method = str(meta.get("method", "") or "")
        path = str(meta.get("path", "") or "")
        for matcher, scope in self._rules:
            if callable(matcher):
                if matcher(ctx):
                    return scope
                continue
            m_method, m_path = matcher
            if method.upper() != m_method.upper():
                continue
            if isinstance(m_path, str):
                if path == m_path:
                    return scope
            elif m_path.search(path):
                return scope
        return None


__all__ = ["DefaultScopePolicy"]
